test/integration/fake_upstream.h

```diff
@@ -130,6 +131,96 @@ class FakeStream : public Http::StreamDecoder,
 
 typedef std::unique_ptr<FakeStream> FakeStreamPtr;
 
+// Encapsulates various state and functionality related to sharing a Connection object across
+// threads. With FakeUpstream fabricated objects, we have a Connection that is associated with a
+// dispatcher on a thread managed by FakeUpstream. We want to be able to safely invoke methods on
+// this object from other threads (e.g. the main test thread) and be able to track connection state
+// (e.g. are we disconnected and the Connection is now possibly deleted). We manage this via a
+// SharedConnectionWrapper that lives from when the Connection is added to the accepted connection
+// queue and then through the lifetime of the Fake{Raw,Http}Connection that manages the Connection
+// through active use.
+class SharedConnectionWrapper : public Network::ConnectionCallbacks {
+public:
+  using DisconnectCallback = std::function<void()>;
+
+  SharedConnectionWrapper(Network::Connection& connection, bool allow_unexpected_disconnects)
+      : connection_(connection), allow_unexpected_disconnects_(allow_unexpected_disconnects) {
+    connection_.addConnectionCallbacks(*this);
+  }
+
+  void addDisconnectCallback(DisconnectCallback callback) {
+    Thread::LockGuard lock(lock_);
+    disconnect_callbacks_.emplace_back(callback);
+  }
+
+  // Network::ConnectionCallbacks
+  void onEvent(Network::ConnectionEvent event) override {
+    // Throughout this entire function, we know that the connection_ cannot disappear, since this
+    // callback is invoked prior to connection_ deferred delete. We also know by locking below, that
+    // elsewhere where we also hold lock_, that the connection cannot disappear inside the locked
+    // scope.
+    Thread::LockGuard lock(lock_);
+    if (event == Network::ConnectionEvent::RemoteClose ||
+        event == Network::ConnectionEvent::LocalClose) {
+      disconnected_ = true;
+      for (const auto& disconnect_callback : disconnect_callbacks_) {
+        disconnect_callback();
+      }
+    }
+  }
+
+  void onAboveWriteBufferHighWatermark() override {}
+  void onBelowWriteBufferLowWatermark() override {}
+
+  bool connected() {
+    Thread::LockGuard lock(lock_);
+    return !disconnected_;
+  }
+
+  // This provides direct access to the underlying connection, but only to const methods.
+  // Stateful connection related methods should happen on the connection's dispatcher via
+  // executeOnDispatcher.
+  // TODO(htuch): This seems a sketchy pattern; even if we're using const methods, there may be
+  // thread safety violations when crossing between the test thread and FakeUpstream thread.
+  Network::Connection& connection() const { return connection_; }
```

Comment:
Can this be made const, or is this comment about const methods wishful thinking?

Reply:
Wishful thinking; even a dispatcher post via the connection is non-const :/

Reply:
Yeah agreed, not sure if this will work in all cases, but we can see what else breaks and fix it.